File: killer_sudoku/solver/engine/rules/cell_solution_elimination.py
"""R1b CellSolutionElimination — eliminate a confirmed digit from all unit peers.

Fires on CELL_SOLVED (emitted by the engine after CELL_DETERMINED).  Receives
cell=(r,c) and hint_digit=d.  Returns Eliminations removing d from all cells
sharing a row, col, or box (cage peers are handled by R3/R4 cage rules).

This is the peer-propagation step of a naked single: once the engine has
committed a cell's solution via CELL_SOLVED, this rule ensures d is removed
from every peer in the same row, column, and 3×3 box.

Separation from NakedSingle (R1a):
  R1a (CELL_DETERMINED) — recognition and placement hint
  R1b (CELL_SOLVED)     — peer candidate elimination

This two-trigger chain makes the data flow explicit: NakedSingle acknowledges
the cell first, then CellSolutionElimination cleans up peers.
"""

from __future__ import annotations

from killer_sudoku.solver.engine.hint import HintResult
from killer_sudoku.solver.engine.rule import RuleContext
from killer_sudoku.solver.engine.types import Elimination, RuleResult, Trigger, UnitKind


class CellSolutionElimination:
    """R1b: eliminate a confirmed digit from all row/col/box peers."""

    name = "CellSolutionElimination"
    priority = 0
    triggers: frozenset[Trigger] = frozenset({Trigger.CELL_SOLVED})
    unit_kinds: frozenset[UnitKind] = frozenset()  # cell-scoped

    def apply(self, ctx: RuleContext) -> RuleResult:
        """Eliminate hint_digit from all row/col/box peers of ctx.cell."""
        assert ctx.cell is not None
        assert ctx.hint_digit is not None
        r, c = ctx.cell
        d = ctx.hint_digit
        elims: list[Elimination] = []
        for uid in ctx.board.cell_unit_ids(r, c):
            unit = ctx.board.units[uid]
            if unit.kind == UnitKind.CAGE:
                continue  # cage peers handled by R3/R4
            for pr, pc in unit.cells:
                if (pr, pc) != (r, c) and d in ctx.board.candidates[pr][pc]:
                    elims.append(Elimination(cell=(pr, pc), digit=d))
        return RuleResult(eliminations=elims)

    def as_hints(
        self, ctx: RuleContext, eliminations: list[Elimination]
    ) -> list[HintResult]:
        """Describe the peer eliminations for a solved cell.

        CellSolutionElimination fires on CELL_SOLVED; ctx.cell is the
        determined cell and ctx.hint_digit is its placed value.
        """
        if not eliminations:
            return []
        assert ctx.cell is not None
        assert ctx.hint_digit is not None
        r, c = ctx.cell
        d = ctx.hint_digit
        peer_labels = ", ".join(
            sorted(f"r{e.cell[0] + 1}c{e.cell[1] + 1}" for e in eliminations)
        )
        return [
            HintResult(
                rule_name=self.name,
                display_name="Naked Single",
                explanation=(
                    f"Cell r{r + 1}c{c + 1} is {d}. "
                    f"Eliminating {d} from peers: {peer_labels}."
                ),
                highlight_cells=frozenset({ctx.cell} | {e.cell for e in eliminations}),
                eliminations=eliminations,
            )
        ]

File: web/src/engine/rules/cellSolutionElimination.ts
/**
 * CellSolutionElimination — R1b: solved cell removes digit from row/col/box peers
 * and from distinct-digit cage peers.
 *
 * Mirrors Python's `killer_sudoku.solver.engine.rules.cell_solution_elimination` module.
 *
 * Fires on CELL_SOLVED. Eliminates hintDigit from every unit peer of ctx.cell,
 * including cells in the same distinct-digit cage (which have the same exclusion
 * property as rows/cols/boxes). Non-distinct cage peers are skipped because those
 * cages allow repeated digits.
 */

import type { HintResult } from '../hint.js';
import type { RuleContext } from '../rule.js';
import {
  Cell,
  Elimination,
  emptyResult,
  RuleResult,
  Trigger,
  UnitKind,
} from '../types.js';
import { cellLabel } from './_labels.js';

export class CellSolutionElimination {
  readonly name = 'CellSolutionElimination';
  readonly displayName = 'Cell Solution Elimination';
  readonly description = `\
Cell Solution Elimination — a placed digit is removed from all peer cells in shared units.

When cell C is assigned digit d, every other cell in C's row, column, box, and distinct-digit cage cannot hold d (each unit must contain each digit at most once, and for distinct-digit cages no digit may repeat within the cage).

Proof: C = d. For any peer P ≠ C sharing a unit U with C: if d ∈ candidates(P), the unit constraint would require two occurrences of d in U — contradiction. Therefore d can be eliminated from every such P.

Non-distinct cage peers are excluded: those cages permit repeated digits, so the cage constraint does not apply.

Guards:
  ctx.cell !== null                engine sets this on CELL_SOLVED
  ctx.hintDigit !== null           the digit being placed
  unit.kind === UnitKind.CAGE → unit.distinctDigits   skip cages that allow repeated digits`.trim();
  readonly priority = 0;
  readonly triggers: ReadonlySet<Trigger> = new Set([Trigger.CELL_SOLVED]);
  readonly unitKinds: ReadonlySet<UnitKind> = new Set();

  apply(ctx: RuleContext): RuleResult {
    if (ctx.cell === null || ctx.hintDigit === null) return emptyResult();
    const [r, c] = ctx.cell;
    const d = ctx.hintDigit;
    const elims: Elimination[] = [];
    for (const uid of ctx.board.cellUnitIds(r, c)) {
      const unit = ctx.board.units[uid]!;
      if (unit.kind === UnitKind.CAGE && !unit.distinctDigits) continue;
      for (const [pr, pc] of unit.cells as Cell[]) {
        if (!(pr === r && pc === c) && ctx.board.cands(pr, pc).has(d))
          elims.push({ cell: [pr, pc] as Cell, digit: d });
      }
    }
    return { ...emptyResult(), eliminations: elims };
  }

  asHints(ctx: RuleContext, eliminations: readonly Elimination[]): HintResult[] {
    if (!eliminations.length || ctx.cell === null || ctx.hintDigit === null) return [];
    const [r, c] = ctx.cell;
    const d = ctx.hintDigit;
    const peerLabels = [...eliminations]
      .sort((a, b) => a.cell[0] - b.cell[0] || a.cell[1] - b.cell[1])
      .map(e => cellLabel(e.cell))
      .join(', ');
    return [{
      ruleName: this.name,
      displayName: 'Naked Single',
      explanation: `Cell ${cellLabel([r, c] as Cell)} is ${d}. Eliminating ${d} from peers: ${peerLabels}.`,
      highlightCells: [[r, c] as Cell, ...eliminations.map(e => e.cell)],
      eliminations: [...eliminations],
      placement: null,
      virtualCageSuggestion: null,
    }];
  }
}
